virtualization/Docker/Dockerfile.dev

```diff
@@ -40,14 +40,6 @@ RUN curl -sL https://deb.nodesource.com/setup_7.x | bash - && \
 # Install tox
 RUN pip3 install --no-cache-dir tox
 
-# Copy over everything required to run tox
```

Comment:
This change causes every code change in hass sources to have tox download everything again (terrible during development in docker). Docker cashes it's layers so it is recommended to download dependencies based on a requirement file before adding the code to the image.

Reply:
@michaelarnauts I made this change based on the docker usage in the `script/test_docker` (here: https://github.com/home-assistant/home-assistant/blob/dev/script/test_docker#L14). When using that script (or the `lint_docker` script in the same folder), the tox cache is on the host, outside of the docker image. The internal cache built by this step in the Dockerfile is totally unused. I see it was actually your commit that did this a year ago (great idea -- it speeds it up a lot! :) https://github.com/home-assistant/home-assistant/pull/5177

Has your workflow changed in a way that makes this tox cache inside the docker image used? Of course, if it is being used, we should not remove it. However, if it's not being used by anyone, then we should remove it, since it adds many minutes to the docker build every time a package version changes

Reply:
You are right that this tox cache that is created during build phase is not used during test_docker and lint_docker since the mounted .tox folder is used there (I already forgot about my changes from a year ago).

The --notest was already in the Dockerfile before I made changes (#2673), so I left it there. I think it was a benefit for running the tests in the container, but since the mapping to .tox this might be obsolete indeed.

It's been a while while I used the development-inside-docker workflow for hass, but I remember there was an issue where on every `script/dev_docker` run, all the pip dependencies had to be downloaded again. Looking at the code, I don't think that the tox cache could fix this.

Reply:
Fortunately, I'm not seeing the issue you're referencing about all the pip dependencies needing to re-download on every run. Using the code from this PR, I'm able to run the `script/test_docker` and `script/dev_docker`, and both run without downloading anything after the first run.

However, I think I see what you're referring to: if the `requirements_all.txt` file changes at all, the docker container will rebuild (due to `COPY requirements_all.txt requirements_all.txt`) and this will require downloading all the the dependencies in the container (dockerfile: `RUN pip3 install --no-cache-dir -r requirements_all.txt`).

Even after the docker container rebuilds, the tox cache should still work. And, after this PR, tox will correctly update the dependencies according to `requirements_all.txt`